go/private/rules/cgo.bzl

```diff
@@ -57,6 +57,40 @@ load(
 
 _CgoCodegen = provider()
 _CgoInfo = provider()
+_CcAlwaysLinkInfo = provider()
+
+_WHOLE_ARCHIVE_FLAGS_BY_GOOS = {
+    goos: ("-Wl,--whole-archive", "-Wl,--no-whole-archive")
+    for goos in GOOS.keys()
+} + {
+    "darwin": ("-Wl,-all_load", ""),
+    "js": ("", ""),
+    "windows": ("", ""),
+}
+
+def _always_link_aspect_impl(target, ctx):
```

Comment:
An aspect will be too expensive for this. It gets applied not only to the `cc_library` we depend on through `cdeps`, but to everything it depends on, which could be a lot of things. Aspects consume a lot of time and memory, and I'd love to get rid of the ones we already have.

It's unfortunate that `CcSkylarkApiProvider` doesn't expose this information. Do you know if there's an open issue to add it? If not, maybe we can file one. It seems like a reasonable thing to add.

Until then, I think it would be reasonable to make `-Wl,--whole-archive` the default for all libraries. This would probably fix #1486 (looks like you've already seen that one).

Reply:
Is there a way to measure this? We build a pretty complex codebase and we didn't feel it was slow?

Reply:
I've opened https://github.com/bazelbuild/bazel/issues/7033

Reply:
You can build with `--profile=file` and then `bazel analyze-profile file`. It will tell you how much time Bazel spends in loading, analysis, execution, etc, but it won't say which lines of code cost the most. I don't think there's a way to do that yet.

cgo currently adds a huge amount of loading / analysis overhead because it declares `cc_library` rules for each configuration that could affect build files. That's separate from the `cdeps` libraries, but in any case, it's quite a bit slower than `go build`. An aspect would slow it further.

Reply:
I'm not sure using `-Wl,--whole-archive` would be the best way really, it may bloat the binaries significantly.

That said, if we could find a way to get the `go.o` file, and use a normal `cc_library` to link it against the rest (bypassing `go link`), we might get all the niceties for free.

Reply:
That may be possible, but it would interfere with tooling. `bazel query` would show your target as a `cc_binary` with a `go_binary` dependency. We have some hacks like this already (cgo), but Bazel API changes have made this less necessary.

Thanks for filing bazelbuild/bazel#7033. I think that's the right path forward. I'll comment there.